modules/fee/handler.go

```diff
@@ -11,76 +11,67 @@ import (
+// Bank is a default location for the fees, but pass anything into
+// the middleware constructor
+var Bank = basecoin.Actor{App: NameFee, Address: []byte("bank")}
+
+// SimpleFeeMiddleware - middleware for fee checking, constant amount
+// It used modules.coin to move the money
+type SimpleFeeMiddleware struct {
+	MinFee    coin.Coin //
+	Collector basecoin.Actor
+	stack.PassOption
 }
 
-// SimpleFeeHandler - checker object for fee checking
-type SimpleFeeHandler struct {
-	AccountChecker
-	MinFee coin.Coins
-	stack.PassOption
+var _ stack.Middleware = SimpleFeeMiddleware{}
+
+// NewSimpleFeeMiddleware returns a fee handler with a fixed minimum fee.
+//
+// If minFee is 0, then the FeeTx is optional
+func NewSimpleFeeMiddleware(minFee coin.Coin, collector basecoin.Actor) SimpleFeeMiddleware {
+	return SimpleFeeMiddleware{
+		MinFee:    minFee,
+		Collector: collector,
+	}
 }
 
 // Name - return the namespace for the fee module
-func (SimpleFeeHandler) Name() string {
+func (SimpleFeeMiddleware) Name() string {
 	return NameFee
 }
 
-var _ stack.Middleware = SimpleFeeHandler{}
-
-// Yes, I know refactor a bit... really too late already
-
 // CheckTx - check the transaction
-func (h SimpleFeeHandler) CheckTx(ctx basecoin.Context, store state.KVStore, tx basecoin.Tx, next basecoin.Checker) (res basecoin.Result, err error) {
-	feeTx, ok := tx.Unwrap().(*Fee)
-	if !ok {
-		return res, errors.ErrInvalidFormat(tx)
-	}
-
-	fees := coin.Coins{feeTx.Fee}
-	if !fees.IsGTE(h.MinFee) {
-		return res, ErrInsufficientFees()
-	}
-
-	if !ctx.HasPermission(feeTx.Payer) {
-		return res, errors.ErrUnauthorized()
-	}
-
-	_, err = h.ChangeAmount(store, feeTx.Payer, fees.Negative())
-	if err != nil {
-		return res, err
-	}
-
-	return basecoin.Result{Log: "Valid tx"}, nil
+func (h SimpleFeeMiddleware) CheckTx(ctx basecoin.Context, store state.KVStore, tx basecoin.Tx, next basecoin.Checker) (res basecoin.Result, err error) {
+	return h.doTx(ctx, store, tx, next.CheckTx)
 }
 
 // DeliverTx - send the fee handler transaction
-func (h SimpleFeeHandler) DeliverTx(ctx basecoin.Context, store state.KVStore, tx basecoin.Tx, next basecoin.Deliver) (res basecoin.Result, err error) {
-	feeTx, ok := tx.Unwrap().(*Fee)
+func (h SimpleFeeMiddleware) DeliverTx(ctx basecoin.Context, store state.KVStore, tx basecoin.Tx, next basecoin.Deliver) (res basecoin.Result, err error) {
+	return h.doTx(ctx, store, tx, next.DeliverTx)
+}
+
+func (h SimpleFeeMiddleware) doTx(ctx basecoin.Context, store state.KVStore, tx basecoin.Tx, next basecoin.CheckerFunc) (res basecoin.Result, err error) {
+	feeTx, ok := tx.Unwrap().(Fee)
 	if !ok {
-		return res, errors.ErrInvalidFormat(tx)
+		// the fee wrapper is not required if there is no minimum
+		if h.MinFee.IsZero() {
+			return next(ctx, store, tx)
+		}
+		return res, errors.ErrInvalidFormat(TypeFees, tx)
 	}
 
-	fees := coin.Coins{feeTx.Fee}
-	if !fees.IsGTE(h.MinFee) {
+	// see if it is big enough...
+	fee := feeTx.Fee
+	if !fee.IsGTE(h.MinFee) {
```

Comment:
Okay - so this call will in fact return an error if the different denom's are contained within `fee` and `h.MinFee` however we are reporting on insufficient fees, which although is technically true, is a bit misleading, because the real issue is the different denom's we should consider adding a denom check right before this.